app/helpers/reports_helper.rb

```diff
@@ -1,2 +1,109 @@
 module ReportsHelper
+  # For taxon_count 'species' rows without a corresponding 'genus' rows,
+  # we create a fake singleton genus containing just that species;
+  # the fake genus IDs start here:
+  FAKE_GENUS_BASE = -1_900_000_000
+
+  def self.validate_names!(counts, lineage_by_tax_id)
+    # This makes up suitable names for missing and blacklisted genera and species. Such
+    # made-up names should be lowercase so they are sorted below proper names.
+    missing_names = Set.new
+    missing_parents = {}
+
+    genus_str = TaxonLineage.level_name(TaxonCount::TAX_LEVEL_GENUS)
+    family_str = TaxonLineage.level_name(TaxonCount::TAX_LEVEL_FAMILY)
+
+    # Iterate through both species and genus.
+    counts.each do |tax_level, tax_counts|
+      tax_counts.each do |tax_id, tax_info|
+        level_str = TaxonLineage.level_name(tax_level)
+
+        if tax_id < 0
+          # Usually -1 means accession number did not resolve to species.
+          # TODO: Can we keep the accession numbers to show in these cases?
+          # NOTE: important to be lowercase for sorting below uppercase valid genuses
+          tax_info[:name] = "all taxa with neither family nor genus classification"
+
+          if tax_id < TaxonLineage::INVALID_CALL_BASE_ID
+            parent_str = tax_level == TaxonCount::TAX_LEVEL_SPECIES ? genus_str : family_str
+            parent_tax_id = tax_level == TaxonCount::TAX_LEVEL_SPECIES ? tax_info[genus_tax_id] : convert_neg_taxid(tax_id)
+            parent_name = fetch_parent_name(tax_level, tax_id, tax_info, lineage_by_tax_id)
+            # If no parent (genus/family) name was found, then use the parent id as the name.
+            unless parent_name
+              missing_parents[parent_tax_id] = tax_id
+              parent_name = "taxon #{parent_tax_id}"
+            end
+            tax_info[:name] = "non-#{level_str}-specific reads in #{parent_str} #{parent_name}"
+            MetricUtil.log_analytics_event("non-#{level_str}-specific reads in #{parent_str} #{parent_name}", nil)
```

Comment:
As I think about lets just move Metric to logs. Metrics need to follow a particular naming structure. 
Can we also complement all logs with pipeline_run_id? Just so we can find it later.